POO-Java/LearnToDoRealJava.md

```diff
@@ -55,43 +61,339 @@ public move(int x, int y) {
+lançar uma exceção, o que fazemos? Voltamos a lançar a exceção? Ignoramos a
+exceção e escrevemos para um log?\
+Efetivamente, uma implementação de `Foo.get_bar()` com base em `.clone()`s
+introduz a necessidade de lidar com exceções. Uma função trivial como um getter
+não devia criar esta responsabilidade no caller.\
+Esta implementação de `get_bar()` é considerada uma
+*[leaky abstraction][LeakyAbstractions]* -- a tentativa de abstração de
+detalhes, neste caso a origem de `bar`, acaba por introduzir novos problemas
+imprevistos. -- o tratamento de exceções.
+
+* **Falsa imutabilidade**: Nós não estamos a fazer `.clone()` porque
+queremos uma cópia, mas sim porque queremos garantir que a nossa variável de
+instância não seja ilegalmente modificada pelo caller, ou, mais corretamente,
+queremos indicar ao caller que não é suposto modificá-la.\
+Imaginemos agora que `Bar` tem a seguinte função:
+    ```java
+    class Bar {
+        public void clear_strings() {
+            this.strings.clear();
+        }
+    }
+    ```
+    Esta função altera o estado interno de `Bar`. Agora, o método
+`Foo.get_bar()` passou a retornar um objeto que permite modificações.\
+Isto não é intuitivo.\
+Estávamos a tentar transmitir ao caller que não é suposto modificar `Bar`, e
+agora estamos a retornar um `Bar` que é modificável. Então agora podemos
+modificar `Bar`? E será que estas alterações são refletidas no estado interno
+de `Foo`?\
+Teríamos de consultar a documentação de `Foo.get_bar()` para perceber o que está
+realmente a acontecer.\
+Não seria tão mais fácil se esta informação estivesse guardada no tipo retornado
+pela `get_bar()`?
+
+* **A mítica interface Cloneable**: Num mundo ~~perfeito~~ razoável, seria de
+esperar que `Cloneable` fosse uma simples interface que contém a declaração de
+um método chamado `.clone()`, *right*? Algo deste género:
+    ```java
+    public interface Cloneable {
+        Object clone();
+    }
+    ```
+    Assim, qualquer objeto que implementava esta interface tinha de definir um
+método `clone()`.
+
+    **NOPE!** Isto faz demasiado sentido.\
+A interface `Cloneable` não declara nenhum método! A sua verdadeira
+implementação é a seguinte:
+    ```java
+    public interface Cloneable {}
+    ```
+    Poderás pensar "*Espera aí, então isso significa que eu posso fazer com que
+a minha class seja `Cloneable`, e não ter nenhum método `.clone()`!*".\
+Yep, estás
+completamente correto! Dizeres que a tua class implementa a interface
+`Cloneable` não significa absolutamente nada, e não traz nenhuma garantia sobre
+o facto de a tua class definir ou não um método `.clone()`! Hurray Java!\
+`Cloneable` é uma interface [completamente *broken*][CloneableBroken]. Não a
+uses.
+
+* **Casts, casts everywhere:** certamente não tão importante como os outros
+argumentos, mas usar `.clone()`s obriga-nos a escrever uma quantidade
+considerável de casts, que acaba por gerar poluição visual no nosso código.
```

Comment:
This is not true, java has return type covariance: https://www.baeldung.com/java-covariant-return-type

Reply:
The cast doesn't happen in the return expression of the clone() method implementation. It happens before the return expression, when deep cloning the fields, e.g. https://github.com/mendess/ResumosMIEI/blob/bd5e08af2ec408850e69a2f39e8adca379050ffb/POO-Java/LearnToDoRealJava.md?plain=1#L134

Reply:
Hmm yeah if you're cloning arraylists like that I guess that happens, the advice given in POO classes is to use the copy ctor `new ArrayList<>(this.strings)`. Which although inconsistent doesn't have this problem.

I just find this remark overall unnecessarily unfair to the pattern.

Reply:
That may be so, but if you look up examples of clone() implementations, most (if not all of them) use the deep clone + cast strategy, occasionally demonstrating the copy constructor delegation strategy.

Moreover, every clone implementation in this file uses the deep clone strategy, not the copy ctor.

I think it's pertinent enough to warrant its own section.

Alternatively, a subsection about copy ctor delegation could be written.